Add Today's Experiment

svidovich opened the pull request:
New changeset for today's experiment:
- Adds README.md to top-level
- Adds new server directory containing server portion
- Adds requirements.txt
- Shifts first project into iso dated folder

svidovich committed:
Changes...
- Adds README.md to top-level
- Adds new server directory containing server portion
- Adds requirements.txt
- Shifts first project into iso dated folder

spencerhcindia commented on 2024-06-10/README.md:
Could we have the README include instructions on how to setup local to spin the API, including how to install package requirements.

svidovich replied:
Done, good callout.

spencerhcindia commented on 2024-06-10/README.md:
```diff
@@ -0,0 +1,14 @@
+## Server portion
+
+To run the server, make sure that you've installed the requirements, and then run
+
```
Perhaps we can also include other OS instructions?

svidovich replied:
I'm thinking that's out of scope here. These instructions should do fine on any *nix, and we're not supporting windows systems at this time.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+
+
+def read_users() -> dict:
+    with open(USERS, "r") as file_handle:
```
What if file doesn't exist?

svidovich replied:
Hey, good eye. Didn't catch this during testing because we hoisted the existence checking logic up into another function. Got this fixed in 7681f8ab1f761e1fa25e85505a7d69288e12b973.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+    address: str | None = None,
+) -> int:
+    users = read_users()
+    with open(USERS, "w") as file_handle:
```
Does this write a new file every time we call this method?

svidovich replied:
That's right. We're going dumb here, where we just read the whole file in, and overwrite it when we're done.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+    del users[user_id]
+
+    with open(USERS, "w") as file_handle:
+        file_handle.truncate(0)
```
Is truncate redundant because opening file with 'w'?

svidovich replied:
Good eye, yeah that's right. I went ahead and removed all of the truncate calls; we don't need those, now.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    with open(USERS, "w") as file_handle:
+        file_handle.truncate(0)
```
Same as line 62

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+            "error": f"You must provide at least one of {', '.join(args_map.keys())}"
+        }
+    data = read_users().values()
+    non_null_args = {
```
Hey can you add a comment explaining what all of this arg_name line is... pls

svidovich replied:
Gotcha, let me know if what I added suits.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+        arg_name: arg for arg_name, arg in args_map.items() if arg is not None
+    }
+    for entry in data:
+        if all(
```
Could you explain how this if statement works?

svidovich replied:
Sure, I added a comment, but I'll be a little more in-depth up here -- 

`all` operates on an iterable of values, a `list` for example. It checks the "truthiness" of every value in that iterable. So

```python
all([True, 1, "the string"])
```
would evaluate to `True`, because none of the entries are "Falsy".

The next part is a little tougher. We know that each `entry` represents a user. `non_null_args` represents our input arguments, filtered to exclude null entries. Then, we get each value corresponding to an input argument from our user.

If each value from the user matches each value from our input arguments, it means that we've found a user that matches our query.

spencerhcindia commented on 2024-06-10/server/main.py:
```diff
@@ -0,0 +1,161 @@
+
+@router.get("/docs")
+def get_documentation() -> list[dict]:
+    """
```
The largest thank for this comment, would love to see more of this or we will fire you, thank you

svidovich replied:
Oh no! I went ahead and updated all of the docstrings. Let me know what you think.

svidovich committed:
README update...
- Include instructions for setting up virtual environment
- More verbose language around running

svidovich committed:
Updates...
- Remove redundant truncate calls for files
- Hoist file existence checking
- Fill out docstrings
- Add comments around terse code

spencerhcindia reviewed (approved):
cool stuff here lgtm
